### PR TITLE
ramips: add support for DLINK DWR-922-E2

### PR DESCRIPTION
**Pull request by squigley**
Replacement PR for this one: https://github.com/openwrt/openwrt/pull/1656

### CODE REVIEW

**Comment by 981213** [comment]
Please follow https://openwrt.org/submitting-patches to squash your commits and update your original PR instead :)

**Comment by cshoredaniel** [comment]
@981213 @squigley had a bit of trouble with that...do you know of any good tutorials for relative git newbies?  I don't although I'm looking to dig some up to help them.

**Comment by bobafetthotmail** [comment]
imho it's not necessary to invest the time to understand git if you only want to get a patch merged.

You only need to be able to follow 2-3 procedures, and if you can't follow them, the procedure has to be expanded so it starts from a more basic starting step.

**Comment by squigley** [comment]
I have used git before, but only for my own code. This is the first time I'm trying to contribute to something else.

It all went wrong when I modified a file which didn't need to be modified. Initially I re-edited the file to manually revert it, but that didn't make sense, I should have just been able to remove it from the commit/PR. That's where the squashing came in, however never having done that before, and being offered 3 different ways to do it, I ended up attempting a mix of these, which just created a disaster.

After apparently resetting to the wrong HEAD, I didn't want to lose all my changes and have to do everything again, so I tried just copying my modified files out of the tree, and then doing pull/fetch/merge all of which gave me cryptic errors about unmergeable files and fast forwards and things I didn't understand. Then when I managed to get it back to a workable state, it would reject my push.

Then I did something else, perhaps another reset, and when I tried to commit my changes, it was presenting me to sign off on someone else's change to a file, so that wasn't right either.

After trying to go through the cherry picking process, which didn't work for me, I then forced a reset, thinking that would just put me right back where I started when I forked the project, but it was only after I made the changes all over again and re-commited and pushed it, that I realised I still had a new branch created, and hence I ended up with a whole new branch submitted, and as I now was well and truly confused and frustrated I figured I would just create this new PR with the correct files, and be done with it.

So basically my problems seem to be that the documentation is lacking to explain the squash process adequately, as it doesn't explain why I am doing the steps/what the result is, or how to go back if the wrong reset level is done, or why you would be presented with someone else's commit message, and that you shouldn't try doing things 3 different ways, or where to find the branch name/version that you want to sync to/be working on.

I know this is supposed to be where we discuss the specific code I have submitted for submission, but at this point I need to know how I can get rid of this unnecessary branch, switch back to the original one, reset to the commit which resulted in the original push/PR request, then how to modify that commit to remove the unnecessary file, and recommit/push.

**Comment by bobafetthotmail** [comment]
>It all went wrong when I modified a file which didn't need to be modified. Initially I re-edited the file to manually revert it, but that didn't make sense, I should have just been able to remove it from the commit/PR. 

You cannot modify a commit like that, you have to delete the commit, as I said deleting the commit does not change the files so you are not losing your work.
Then you commit again, when you do **git add -i** you only select the files you want to have in that commit. It does not matter that you actually changed other files too. If they are not selected when you do the **git add -i** interface then it will not end in the commit or the PR.

At the end, **git push -f** to send it over to github.

> why I am doing the steps/what the result is,

This is a bit generic, will make the tutorial an actual real-life example with true output and see if it's better.

>or how to go back if the wrong reset level is done, 

Deleting commits like that is permanent. But as long as you did not push the changes to github with **git push** you can just delete your local git folder and clone again from your fork of the openwrt github repo.
Your openwrt fork is here on github https://github.com/squigley/openwrt

In general, if you manage to screw up anything you can just manually copy the files you want to save somewhere safe, then delete the whole local git folder and download again from your own git repo, **git checkout branch-name** to go in the same branch you were in before, then paste your modified files in the right places again. Already having a backup of the local git folder is faster.

It's crude but works for me. I don't feel like delving too deep in git knowledge when plain "restoring a backup" works fine.

>where to find the branch name/version that you want to sync to/be working on.

the github web interface shows the branch name assigned to this PR in the line under the title in the PR page (in this same page we are using)

 _squigley wants to merge 3 commits into openwrt:master from squigley:pr-new-branch_ 

any change you do to the "pr-new-branch" branch and then git push will end up here. 

Your other PR has this line

 _squigley wants to merge 6 commits into openwrt:master from squigley:DlinkDWR922E2_ 

Which means that the branch connected to the PR is DlinkDWR922E2

By looking at the github page of your openwrt fork here https://github.com/squigley/openwrt I can see that (click on "Branch: **master**" button) that there you still have both branches in on your fork.

When you are in your local git folder and ytou want to see all branches you have available you have to write (as said at the beginning of the procedure)

**git branch -a**

and it will list all branches available. The branch you are currently using has an asterisk on its left.

Change the branch you are working on with **git checkout branch-name**

>how I can switch back to the original branch

write

**git checkout DlinkDWR922E2**

>how I can get rid of this unnecessary branch, 

write

**git branch --delete --force pr-new-branch**

and this branch will be deleted.

You will have to close the PR from the Github web interface later.

>reset to the commit which resulted in the original push/PR request

If you just want to delete the last commit (i.e. **Merge branch 'DlinkDWR922E2' of https://github.com/squigley/openwrt i…**  ) which seems to be doing weird things for a device support PR, just do a **git reset HEAD~1**. The last commit will be deleted. You will still have that file modified so yours is a "dirty" branch but it's irrelevant, once the PR is merged you will delete your branch anyway.

If that's all you want to do, then

**git push -f**

To send the changes over to github.

If you want to do something more involved than just nuking the last commit, let's make an example.

Looking at your commit history there https://github.com/squigley/openwrt/commits/DlinkDWR922E2

let's say you want to fix "**Added defautl LED configurationa, modified LED naming, added licence …** "

which is the third commit if we start to count from the last commit in that branch.

Copy all the commit title and descriptions of these three commits. We are going to delete the commits, so you will need to write them again later.
Actual files won't be affected by this.

write a
**git reset HEAD~3**

and now if you write a **git log** you should see that the last commit in your list is now "**scripts: rstrip.sh: allow rpath with sole $ORIGIN token**"

Now you can go and write **git add -i** to select the files you want to include in your commit.
This opens up an interactive interface where you can select what files you want to include in the commit and what you don't want to. There are other ways to do it too, but I do it this way.

If you want to add modified files, press the 2 key and enter. Then you can select all the changed files that you want to include in this specific commit by writing their line number and pressing enter. You can deselect files too. For example if you want to remove the file 1 from the changed file list for the commit write "-1" and press enter.
Press enter again to exit and return to the "main menu" of the interface.
If you added new files to the source tree press the 4 key and enter. It works the same as above, add the new files.
When you are done, press 7 and then enter to exit the interface.
This is the documentation of that mode if you are curious https://git-scm.com/docs/git-add#_interactive_mode

Now you can commit with **git commit --signoff** and add the title and description of the commit you saved before deleting the commit. It will complain that you have "untracked changes", this is ok.

Repeat this procedure for other commits. 

Once you are done, write

**git push -f**

and it will send the changes over to github, updating these two PRs

**Comment by mkresin** [comment]
I fixed the remaining issues and merged the result.
